### PR TITLE
Feature/template simplification

### LINKED ISSUE #13: Simplify the existing CW20 and CW721 templates
## Description

The templates for the [CW20](https://github.com/archway-network/archway-templates/tree/7f0f16306c7ca647bc883f5080d08bab4e7c7e04/cw20/base) and [CW721 off-chain](https://github.com/archway-network/archway-templates/tree/7f0f16306c7ca647bc883f5080d08bab4e7c7e04/cw721/off-chain-metadata) contracts are not optimal since they are just copies of the code that is already available as crates in CosmWasm. We should clean up the code of both to use the crates instead.

## Proposed solution

- [ ] Clean up the CW20 base code and use the [cw20-base](https://crates.io/crates/cw20-base) crate.
- [ ] Remove the CW721 with off-chain metadata, keep only the CW721 with on-chain metadata, and change the [available template options](https://github.com/archway-network/archway-cli/blob/4c65e5101a6bca1fec47079ba9aeb64d2447cd9d/src/commands/new.js#L15-L20) in the CLI.
- [ ] Update the package dependencies where necessary.

### PR DESCRIPTION
**Pull request by gachouchani1999**
Simplification of templates to follow cw721 onchain metdata template. 
Removal of cw721 off-chain metadata 

Closes #13

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,10 +2,13 @@
 
 Archway Smart Contract template projects used by the [Archway Developer CLI](https://github.com/archway-network/archway-cli).
 
-## List of templates
+## List of templatesa
```
Comment (santanaluiz) [inline README.md:5]: typo

File: .idea/vcs.xml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (santanaluiz) [inline .idea/vcs.xml:1]: can you remove all those `.idea` files? maybe add the directory to the `gitignore`

File: cw20/base/.gitignore [hunk 3]
```diff
@@ -1,6 +1,5 @@
 # Build results
 /target
-/artifacts
```
Comment (santanaluiz) [inline cw20/base/.gitignore:3]: is it expected to remove this?

File: cw20/base/NOTICE [hunk 4]
```diff
@@ -1,4 +1,4 @@
-Copyright {{ "now" | date: "%Y" }} {{authors}}
+Copyright 2022 georgeschouchani <gachouchani@ndu.edu.lb>
```
Comment (santanaluiz) [inline cw20/base/NOTICE:1]: is that correct?

File: cw20/base/README.md [hunk 5]
```diff
@@ -25,8 +24,25 @@ cargo install cargo-run-script
...
+
+```sh
+cargo generate --git https://github.com/CosmWasm/cw-template.git --branch <version> --name PROJECT_NAME
+````
+
+Example:
+
 ```sh
-cargo generate --git archway-network/archway-templates.git --name PROJECT_NAME default
+cargo generate --git https://github.com/CosmWasm/cw-template.git --branch 0.16 --name PROJECT_NAME
```
Comment (santanaluiz) [inline cw20/base/README.md:45]: shouldn't it be pointing to this repo?

File: cw20/base/README.md [hunk 6]
```diff
@@ -1,21 +1,20 @@
-# Archway Network Starter Pack
+# CosmWasm Starter Pack
```
Comment (santanaluiz) [inline cw20/base/README.md:1]: is this change correct?

File: cw20/base/README.md [hunk 7]
```diff
@@ -65,7 +81,7 @@ running `cargo check` or `cargo unit-test`.
 
 Once you have your custom repo, you should check out [Developing](./Developing.md) to explain
 more on how to run tests and develop code. Or go through the
-[online tutorial](https://docs.archway.io/docs/create/guides/my-first-dapp/start) to get a better feel
+[online tutorial](https://docs.cosmwasm.com/) to get a better feel
```
Comment (santanaluiz) [inline cw20/base/README.md:84]: is this change expected?

File: cw20/base/README.md [hunk 8]
```diff
@@ -82,9 +98,9 @@ proper description in the README.
 [Gitpod](https://www.gitpod.io/) container-based development platform will be enabled on your project by default.
 
 Workspace contains:
-
-- **rust**: for builds
-- [wasmd](https://github.com/CosmWasm/wasmd): for local node setup and client
-- **jq**: shell JSON manipulation tool
+ - **rust**: for builds
+ - [wasmd](https://github.com/CosmWasm/wasmd): for local node setup and client
```
Comment (santanaluiz) [inline cw20/base/README.md:102]: shouldn't we refer to `archwayd` or docker?

**Review by santanaluiz — CHANGES_REQUESTED** [review]

File: cw20/base/schema/query_msg.json [hunk 9]
```diff
@@ -3,162 +3,12 @@
   "title": "QueryMsg",
   "anyOf": [
     {
-      "description": "Returns the current balance of the given address, 0 if unset. Return type: BalanceResponse.",
       "type": "object",
       "required": [
-        "balance"
+        "get_count"
```
Comment (santanaluiz) [inline cw20/base/schema/query_msg.json:8]: can you check if this change impacts anything in the `docs`?
or the dapp-examples repo

Comment (gachouchani1999) [reply cw20/base/schema/query_msg.json:8]: Issue here is that we cannot update schema for contract so will copy the ones of cw20_base for the time being
```  invalid character `{` in package name: `{{project-name}}`, the first character must be a Unicode XID start character (most letters or `_`)``` Since it is a template

Comment (gachouchani1999) [reply cw20/base/schema/query_msg.json:8]: or possibly would be best to remove schema completely as it is only a template

File: cw20/base-lib/examples/schema.rs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+use {{crate_name}}::msg::{ExecuteMsg, InstantiateMsg, QueryMsg};
+
+fn main() {
+    let mut out_dir = current_dir().unwrap();
+    out_dir.push("schema");
+    create_dir_all(&out_dir).unwrap();
+    remove_schemas(&out_dir).unwrap();
+
+    export_schema(&sxchema_for!(InstantiateMsg), &out_dir);
```
Comment (aelesbao) [inline cw20/base-lib/examples/schema.rs:14]: Typo

File: cw20/base-lib/examples/schema.rs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+use std::env::current_dir;
+use std::fs::create_dir_all;
+
+use cosmwasm_schema::{export_schema, remove_schemas, schema_for};
+
+use {{crate_name}}::msg::{ExecuteMsg, InstantiateMsg, QueryMsg};
```
Comment (aelesbao) [inline cw20/base-lib/examples/schema.rs:6]: Those structs do not exist in this package

File: .github/scripts/test_generate.sh [hunk 12]
```diff
@@ -35,8 +36,6 @@ function test-template() {
       # Debug builds first to fail fast
       echo "Running unit tests ..."
       cargo unit-test
-      echo "Creating schema ..."
-      cargo schema
```
Comment (aelesbao) [inline .github/scripts/test_generate.sh:39]: Please revert. You shouldn't remove this test.

File: .github/workflows/check.yml [hunk 13]
```diff
@@ -1,4 +1,4 @@
-on: [push, pull_request]
```
Comment (aelesbao) [inline .github/workflows/check.yml:1]: The importance of keeping the `pull_request` trigger is testing if the merge to the target branch will succeed. Please, try adding it back.

**Review by aelesbao — CHANGES_REQUESTED** [review]
Please, rename the `cw20/base-lib` folder back to `cw20/base`. Otherwise, the CLI will break when people try to generate a project from a CW20 template.

File: .github/scripts/test_generate.sh [hunk 14]
```diff
@@ -16,7 +16,8 @@ function test-template() {
   echo "#######################################"
   (
     GIT_BRANCH=$(git -C "$REPO_ROOT" branch --show-current)
-
+    echo "Updating rust and cargo ..."
+    rustup update
```
Comment (aelesbao) [inline .github/scripts/test_generate.sh:20]: Please, remove this from inside the loop

File: .github/workflows/check.yml (hunk 13 above)
Comment (gachouchani1999) [reply .github/workflows/check.yml:1]: Was removed since it does not exist in main

Comment (aelesbao) [reply .github/workflows/check.yml]: Yes, I saw when you removed this trigger in #14, but I recommend trying to add it back to increase the test coverage.

File: cw20/base/.github/workflows/Basic.yml [hunk 15]
```diff
@@ -64,20 +64,12 @@ jobs:
...
-        uses: tj-actions/verify-changed-files@v8
-        id: verify-schema
-        with:
-          files: schema/.*\.json
-
-      - name: Display changed schemas
-        if: steps.verify-schema.outputs.files_changed == 'true'
-        run: |
-          echo "The schema files are not in sync with the repository. Please, run 'cargo schema' to generate them again and commit the changes."
-          exit 1
```
Comment (aelesbao) [inline cw20/base/.github/workflows/Basic.yml:83]: Please revert this change. I added the message to clarify how to fix it for the users.

File: cw20/base/.gitignore [hunk 16]
```diff
@@ -1,16 +1,12 @@
-# Build results
-/target
-/artifacts
```
Comment (aelesbao) [inline cw20/base/.gitignore:3]: Please, keep `/artifacts` ignored.

File: cw20/base/Cargo.toml [hunk 17]
```diff
@@ -6,7 +6,8 @@ edition = "2018"
 
 exclude = [
   # Those files are rust-optimizer artifacts. You might want to commit them for convenience but they should not be part of the source code publication.
-  "artifacts/*",
```
Comment (aelesbao) [inline cw20/base/Cargo.toml:9]: Please revert it; we store the produced wasm files in the `artifacts` folder.

File: cw20/base/Cargo.toml [hunk 18]
```diff
@@ -33,22 +34,22 @@ library = []
...
-cw-storage-plus = "0.12"
-cw0 = "0.10"
-cw2 = "0.12"
-cw20 = "0.12"
-schemars = "0.8"
-serde = { version = "1.0", default-features = false, features = ["derive"] }
-thiserror = "1.0"
+cosmwasm-std = "1.0.0"
+cosmwasm-storage = "1.0.0"
+cw-storage-plus = "0.13.4"
```
Comment (aelesbao) [inline cw20/base/Cargo.toml:45]: It's usually better to use only `0.13` instead of specifying the specific patch number. This way, if the crate has a new patch deployed, the project doesn't need to be updated again - it will automatically update to `0.13.5` when you run `cargo install` for the first time or by running `cargo update`. This comment applies to all other crates below and in other `Cargo.toml` files.

**Review by aelesbao — CHANGES_REQUESTED** [review]

File: cw20/base/Cargo.toml (hunk 18 above)
Comment (aelesbao) [reply cw20/base/Cargo.toml]: Please, check this comment as it's not resolved yet

File: cw20/base/.github/workflows/Basic.yml [hunk 19]
```diff
@@ -37,6 +37,18 @@ jobs:
...
+        id: verify-schema
+        with:
+          files: schema/.*\.json
+
+      - name: Display changed schemas
+        if: steps.verify-schema.outputs.files_changed == 'true'
+        run: |
+          echo "The schema files are not in sync with the repository. Please, run 'cargo schema' to generate them again and commit the changes."
+          exit 1
+
```
Comment (aelesbao) [inline cw20/base/.github/workflows/Basic.yml:51]: Please, check the diff from the previous version. You moved the code to the wrong place.

File: cw721/on-chain-metadata/Cargo.toml [hunk 20]
```diff
@@ -41,10 +41,10 @@ optimize = """docker run --rm -v "$(pwd)":/code \
...
 [dependencies]
 cosmwasm-std = "1.0.0-beta"
-cw2 = "0.11"
-cw721 = "0.11"
-cw721-base = { version = "0.11", features = ["library"] }
-schemars = "0.8"
+cw2 = "0.13.4"
+cw721 = "0.13.2"
+cw721-base = { version = "0.13.2", features = ["library"] }
+schemars = "0.8.10"
```
Comment (aelesbao) [inline cw721/on-chain-metadata/Cargo.toml:47]: Please, check the unresolved comment from the previous review regarding versions with `0.13` vs. `0.13.x`.

**Review by aelesbao — CHANGES_REQUESTED** [review]
Before merging this code, we should open a PR in the CLI to update the templates list, removing the CW721 off-chain option.

File: cw721/on-chain-metadata/Cargo.toml (hunk 20 above)
Comment (gachouchani1999) [reply cw721/on-chain-metadata/Cargo.toml:47]: I did change it but maybe your push reverted them, changing again

Comment (aelesbao) [reply cw721/on-chain-metadata/Cargo.toml:47]: There was nothing extra on the branch. Otherwise, GitHub wouldn't allow me to push without a `--force` push.

Comment (gachouchani1999) [reply cw721/on-chain-metadata/Cargo.toml]: Turns out it was for the on-chain metadata and only changed it for `cw20`

File: cw20/base/.github/workflows/Basic.yml (hunk 15 above)
Comment (aelesbao) [reply cw20/base/.github/workflows/Basic.yml:83]: @gachouchani1999 this is still not resolved. You moved this exact code to the wrong place in the file. Please, recheck it.

**Review by aelesbao — CHANGES_REQUESTED** [review]
